src/decoder.ts

```diff
@@ -139,201 +136,262 @@ export class InvalidErrorError extends Error {
+    const val = (this.peek() as Kind) === Kind.Any;
+    if (val) {
+      this.#pos += 1;
+    }
+    return val;
+  }
+
+  boolean(): boolean {
+    const kind = this.pop() as Kind;
+    if (kind !== Kind.Boolean) {
+      throw new InvalidBooleanError();
+    }
+    return this.pop() === BOOLEAN_TRUE;
+  }
+
+  uint8(): number {
+    const kind = this.pop() as Kind;
+    if (kind !== Kind.Uint8) {
+      throw new InvalidUint8Error();
+    }
+    const dataView = new DataView(this.buf.buffer, this.#pos, 1);
+    const val = dataView.getUint8(0);
+    this.#pos += 1;
+    return val;
+  }
+
+  uint16(): number {
+    const kind = this.pop() as Kind;
+    if (kind !== Kind.Uint16) {
+      throw new InvalidUint16Error();
+    }
+
+    let num = 0;
+    let shift = 0;
+    for (let i = 1; i < MAXLEN16 + 1; i += 1) {
+      const b = this.pop();
+      if (b < CONTINUATION) {
+        return (num | (b << shift)) >>> 0;
+      }
+      num |= b & ((CONTINUATION - 1) << shift);
+      shift += 7;
+    }
+    return num;
+  }
+
+  uint32(): number {
+    const kind = this.pop() as Kind;
+    if (kind !== Kind.Uint32) {
+      throw new InvalidUint32Error();
+    }
+
+    let num = 0n;
+    let shift = 0n;
+    for (let i = 1; i < MAXLEN32 + 1; i += 1) {
+      const b = this.pop();
+      if (b < BigInt(CONTINUATION)) {
+        num += BigInt(b) << shift;
+        return Number(num);
+      }
+      num += BigInt(b & (CONTINUATION - 1)) << shift;
+      shift += 7n;
+    }
+    return Number(num);
+  }
+
+  uint64(): bigint {
+    const kind = this.pop() as Kind;
+    if (kind !== Kind.Uint64) {
+      throw new InvalidUint64Error();
+    }
+
+    let num = 0n;
+    let shift = 0n;
+    for (let i = 1; i < MAXLEN64 + 1; i += 1) {
+      const b = this.pop();
+      if (b < BigInt(CONTINUATION)) {
+        num += BigInt(b) << shift;
+        return num;
+      }
+      num += BigInt(b & (CONTINUATION - 1)) << shift;
+      shift += 7n;
+    }
+    return num;
+  }
+
+  int32(): number {
+    const kind = this.pop() as Kind;
+    if (kind !== Kind.Int32) {
+      throw new InvalidInt32Error();
+    }
+
+    let num = 0n;
+    let shift = 0n;
+    for (let i = 1; i < MAXLEN32 + 1; i += 1) {
+      const b = this.pop();
+      if (b < BigInt(CONTINUATION)) {
+        num += BigInt(b) << shift;
+        // two's complement
+        num = num % 2n === 0n ? num / 2n : -(num + 1n) / 2n;
+        return Number(num);
+      }
+      num += BigInt(b & (CONTINUATION - 1)) << shift;
+      shift += 7n;
+    }
+    return Number(num);
+  }
+
+  int64(): bigint {
+    const kind = this.pop() as Kind;
+    if (kind !== Kind.Int64) {
+      throw new InvalidInt64Error();
+    }
+
+    let num = 0n;
+    let shift = 0n;
+    for (let i = 1; i < MAXLEN64 + 1; i += 1) {
+      const b = this.pop();
+      if (b < BigInt(CONTINUATION)) {
+        num += BigInt(b) << shift;
+        // two's complement
+        num = num % 2n === 0n ? num / 2n : -(num + 1n) / 2n;
```

Comment:
Maybe we can merge the bitwise ops here with `uint32`, `unit64`, `int32` etc. so that we don't need to duplicate them?